File: packages/sdk/src/provenance.ts
/**
 * Provenance store — SQLite-backed trace event persistence.
 */

import Database from 'better-sqlite3';
import type { TraceEvent, TraceQueryOpts, TraceEventStats } from './types.js';
import type { Logger } from './logger.js';

interface TraceRow {
  timestamp: string;
  session_id: string;
  trace_id: string;
  parent_trace_id: string | null;
  tenant_id: string;
  actor: string;
  operation: string;
  inputs: string | null;
  outputs: string | null;
  tokens_in: number;
  tokens_out: number;
  cost_usd: number;
  score: number | null;
  feedback: string | null;
  provenance: string;
}

export class ProvenanceStore {
  private db: Database.Database;
  private log: Logger;

  constructor(db: Database.Database, logger: Logger) {
    this.db = db;
    this.log = logger;
    this.initSchema();
  }

  private initSchema(): void {
    this.db.exec(`
      CREATE TABLE IF NOT EXISTS trace_events (
        trace_id TEXT PRIMARY KEY,
        timestamp TEXT NOT NULL,
        session_id TEXT NOT NULL,
        parent_trace_id TEXT,
        tenant_id TEXT NOT NULL,
        actor TEXT NOT NULL,
        operation TEXT NOT NULL,
        inputs TEXT,
        outputs TEXT,
        tokens_in INTEGER NOT NULL DEFAULT 0,
        tokens_out INTEGER NOT NULL DEFAULT 0,
        cost_usd REAL NOT NULL DEFAULT 0,
        score REAL,
        feedback TEXT,
        provenance TEXT NOT NULL
      );
      CREATE INDEX IF NOT EXISTS idx_trace_session ON trace_events(session_id);
      CREATE INDEX IF NOT EXISTS idx_trace_tenant ON trace_events(tenant_id);
      CREATE INDEX IF NOT EXISTS idx_trace_operation ON trace_events(operation);
      CREATE INDEX IF NOT EXISTS idx_trace_timestamp ON trace_events(timestamp);
    `);
  }

  record(event: TraceEvent): void {
    this.db
      .prepare(
        `
      INSERT OR REPLACE INTO trace_events
        (trace_id, timestamp, session_id, parent_trace_id, tenant_id, actor,
         operation, inputs, outputs, tokens_in, tokens_out, cost_usd,
         score, feedback, provenance)
      VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
    `
      )
      .run(
        event.traceId,
        event.timestamp,
        event.sessionId,
        event.parentTraceId ?? null,
        event.tenantId,
        JSON.stringify(event.actor),
        event.operation,
        event.inputs != null ? JSON.stringify(event.inputs) : null,
        event.outputs != null ? JSON.stringify(event.outputs) : null,
        event.tokensIn,
        event.tokensOut,
        event.costUsd,
        event.score ?? null,
        event.feedback ?? null,
        JSON.stringify(event.provenance)
      );
  }

  get(traceId: string): TraceEvent | undefined {
    const row = this.db
      .prepare('SELECT * FROM trace_events WHERE trace_id = ?')
      .get(traceId) as TraceRow | undefined;
    return row ? this.toEvent(row) : undefined;
  }

  query(opts: TraceQueryOpts = {}): TraceEvent[] {
    const conditions: string[] = [];
    const params: unknown[] = [];

    if (opts.sessionId) {
      conditions.push('session_id = ?');
      params.push(opts.sessionId);
    }
    if (opts.tenantId) {
      conditions.push('tenant_id = ?');
      params.push(opts.tenantId);
    }
    if (opts.operation) {
      conditions.push('operation = ?');
      params.push(opts.operation);
    }
    if (opts.since) {
      conditions.push('timestamp >= ?');
      params.push(opts.since);
    }

    const where =
      conditions.length > 0 ? `WHERE ${conditions.join(' AND ')}` : '';
    const limit = opts.limit ?? 100;

    const rows = this.db
      .prepare(
        `SELECT * FROM trace_events ${where} ORDER BY timestamp DESC LIMIT ?`
      )
      .all(...params, limit) as TraceRow[];

    return rows.map((r) => this.toEvent(r));
  }

  supersede(traceId: string, supersededBy: string): void {
    const row = this.db
      .prepare('SELECT provenance FROM trace_events WHERE trace_id = ?')
      .get(traceId) as { provenance: string } | undefined;
    if (!row) return;

    const prov = JSON.parse(row.provenance) as TraceEvent['provenance'];
    prov.supersededBy = supersededBy;
    this.db
      .prepare('UPDATE trace_events SET provenance = ? WHERE trace_id = ?')
      .run(JSON.stringify(prov), traceId);
  }

  getLineage(traceId: string): TraceEvent[] {
    const chain: TraceEvent[] = [];
    let current = this.get(traceId);
    while (current) {
      chain.unshift(current);
      if (!current.parentTraceId) break;
      current = this.get(current.parentTraceId);
    }
    return chain;
  }

  getStats(tenantId?: string): TraceEventStats {
    const where = tenantId ? 'WHERE tenant_id = ?' : '';
    const params = tenantId ? [tenantId] : [];

    const row = this.db
      .prepare(
        `
      SELECT COUNT(*) as total,
             COALESCE(SUM(tokens_in), 0) as tokens_in,
             COALESCE(SUM(tokens_out), 0) as tokens_out,
             COALESCE(SUM(cost_usd), 0) as cost_usd,
             COALESCE(AVG(json_extract(provenance, '$.confidence')), 0) as avg_confidence
      FROM trace_events ${where}
    `
      )
      .get(...params) as {
      total: number;
      tokens_in: number;
      tokens_out: number;
      cost_usd: number;
      avg_confidence: number;
    };

    return {
      totalEvents: row.total,
      totalTokensIn: row.tokens_in,
      totalTokensOut: row.tokens_out,
      totalCostUsd: row.cost_usd,
      avgConfidence: row.avg_confidence,
    };
  }

  private toEvent(row: TraceRow): TraceEvent {
    const event: TraceEvent = {
      timestamp: row.timestamp,
      sessionId: row.session_id,
      traceId: row.trace_id,
      tenantId: row.tenant_id,
      actor: JSON.parse(row.actor),
      operation: row.operation,
      inputs: row.inputs ? JSON.parse(row.inputs) : null,
      outputs: row.outputs ? JSON.parse(row.outputs) : null,
      tokensIn: row.tokens_in,
      tokensOut: row.tokens_out,
      costUsd: row.cost_usd,
      provenance: JSON.parse(row.provenance),
    };
    if (row.parent_trace_id) event.parentTraceId = row.parent_trace_id;
    if (row.score != null) event.score = row.score;
    if (row.feedback) event.feedback = row.feedback;
    return event;
  }
}

File: packages/python-sdk/stackmemory/provenance.py
"""ASI-shaped trace events with provenance tracking."""

from __future__ import annotations

import json
import sqlite3
import uuid
from dataclasses import dataclass, field, asdict
from datetime import datetime


@dataclass
class SourceRef:
    type: str  # tool, user, agent, ingestion, cache
    id: str
    label: str | None = None


@dataclass
class Provenance:
    sources: list[SourceRef] = field(default_factory=list)
    derivation: list[str] = field(default_factory=list)
    confidence: float = 1.0
    superseded_by: str | None = None


@dataclass
class Actor:
    host: str = "unknown"
    agent: str = "stackmemory"
    user: str = "anonymous"


@dataclass
class TraceEvent:
    timestamp: str = ""
    session_id: str = ""
    trace_id: str = ""
    tenant_id: str = "local"
    actor: Actor = field(default_factory=Actor)
    operation: str = ""
    inputs: dict = field(default_factory=dict)
    outputs: dict = field(default_factory=dict)
    tokens_in: int = 0
    tokens_out: int = 0
    cost_usd: float = 0.0
    duration_ms: int = 0
    score: float | None = None
    feedback: str | None = None
    provenance: Provenance = field(default_factory=Provenance)
    error: str | None = None
    tags: list[str] | None = None
    parent_trace_id: str | None = None


class ProvenanceStore:
    """SQLite-backed ASI-shaped trace event store."""

    def __init__(self, db: sqlite3.Connection) -> None:
        self._db = db
        self._init_schema()

    def _init_schema(self) -> None:
        self._db.executescript("""
            CREATE TABLE IF NOT EXISTS trace_events (
                id TEXT PRIMARY KEY,
                timestamp TEXT NOT NULL,
                session_id TEXT NOT NULL,
                trace_id TEXT NOT NULL,
                parent_trace_id TEXT,
                tenant_id TEXT NOT NULL DEFAULT 'local',
                actor_host TEXT NOT NULL DEFAULT 'unknown',
                actor_agent TEXT NOT NULL DEFAULT 'stackmemory',
                actor_user TEXT NOT NULL DEFAULT 'anonymous',
                operation TEXT NOT NULL,
                inputs TEXT NOT NULL DEFAULT '{}',
                outputs TEXT NOT NULL DEFAULT '{}',
                tokens_in INTEGER NOT NULL DEFAULT 0,
                tokens_out INTEGER NOT NULL DEFAULT 0,
                cost_usd REAL NOT NULL DEFAULT 0,
                duration_ms INTEGER NOT NULL DEFAULT 0,
                score REAL,
                feedback TEXT,
                provenance TEXT NOT NULL DEFAULT '{"sources":[],"derivation":[],"confidence":1}',
                error TEXT,
                tags TEXT
            );
            CREATE INDEX IF NOT EXISTS idx_te_session ON trace_events(session_id);
            CREATE INDEX IF NOT EXISTS idx_te_operation ON trace_events(operation);
            CREATE INDEX IF NOT EXISTS idx_te_timestamp ON trace_events(timestamp);
        """)

    def record(self, event: TraceEvent) -> str:
        """Record a trace event. Returns the generated ID."""
        event_id = str(uuid.uuid4())
        if not event.timestamp:
            event.timestamp = datetime.now(tz=__import__('datetime').timezone.utc).isoformat()

        prov = json.dumps({
            "sources": [asdict(s) for s in event.provenance.sources],
            "derivation": event.provenance.derivation,
            "confidence": event.provenance.confidence,
            "superseded_by": event.provenance.superseded_by,
        })

        self._db.execute(
            """INSERT INTO trace_events (
                id, timestamp, session_id, trace_id, parent_trace_id, tenant_id,
                actor_host, actor_agent, actor_user,
                operation, inputs, outputs,
                tokens_in, tokens_out, cost_usd, duration_ms,
                score, feedback, provenance, error, tags
            ) VALUES (?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?)""",
            (
                event_id, event.timestamp, event.session_id, event.trace_id,
                event.parent_trace_id, event.tenant_id,
                event.actor.host, event.actor.agent, event.actor.user,
                event.operation, json.dumps(event.inputs), json.dumps(event.outputs),
                event.tokens_in, event.tokens_out, event.cost_usd, event.duration_ms,
                event.score, event.feedback, prov, event.error,
                json.dumps(event.tags) if event.tags else None,
            ),
        )
        self._db.commit()
        return event_id

    def annotate(self, event_id: str, score: float | None = None, feedback: str | None = None) -> bool:
        """Add score/feedback to an existing event."""
        updates, params = [], []
        if score is not None:
            updates.append("score = ?")
            params.append(score)
        if feedback is not None:
            updates.append("feedback = ?")
            params.append(feedback)
        if not updates:
            return False

        params.append(event_id)
        cur = self._db.execute(
            f"UPDATE trace_events SET {', '.join(updates)} WHERE id = ?", params
        )
        self._db.commit()
        return cur.rowcount > 0

    def query(
        self,
        session_id: str | None = None,
        operation: str | None = None,
        min_score: float | None = None,
        limit: int = 100,
    ) -> list[dict]:
        """Query events with filters."""
        conditions, params = [], []
        if session_id:
            conditions.append("session_id = ?")
            params.append(session_id)
        if operation:
            conditions.append("operation = ?")
            params.append(operation)
        if min_score is not None:
            conditions.append("score >= ?")
            params.append(min_score)

        where = "WHERE " + " AND ".join(conditions) if conditions else ""
        params.append(limit)
        rows = self._db.execute(
            f"SELECT * FROM trace_events {where} ORDER BY timestamp DESC LIMIT ?",
            params,
        ).fetchall()

        cols = [d[0] for d in self._db.execute("SELECT * FROM trace_events LIMIT 0").description]
        return [dict(zip(cols, row)) for row in rows]

    def get_stats(self, session_id: str | None = None) -> dict:
        """Aggregate statistics."""
        where, params = "", []
        if session_id:
            where = "WHERE session_id = ?"
            params.append(session_id)

        row = self._db.execute(f"""
            SELECT COUNT(*), COALESCE(SUM(tokens_in),0), COALESCE(SUM(tokens_out),0),
                   COALESCE(SUM(cost_usd),0), AVG(score),
                   SUM(CASE WHEN feedback IS NOT NULL THEN 1 ELSE 0 END),
                   SUM(CASE WHEN error IS NOT NULL THEN 1 ELSE 0 END)
            FROM trace_events {where}
        """, params).fetchone()

        return {
            "total_events": row[0],
            "total_tokens_in": row[1],
            "total_tokens_out": row[2],
            "total_cost_usd": row[3],
            "avg_score": row[4],
            "events_with_feedback": row[5],
            "events_with_errors": row[6],
        }
